### PR TITLE
Expand Submitter to HTCondor

### PR DESCRIPTION
**Pull request by anmalara**
Addressing #42.

- [x] Created TimeFormat class. It stores the days, hours, minutes, seconds and performs comparisons.
- [x] Created  ClusterSpecificSettings. It's a class that stores information related to each supported Cluster. It also provides a method to get the closest upper limit for the time allowed by the Cluster (given an input time).
- [X] simplified utils func. 
- [x] Create CondorBase class. Uses python bindings of htcondor, creates and stores the Job Info, submits a single job or a list of jobs to the clusters, allow interactions with the Cluster.
- [x] Exented Submitter and submit.py to support HTCondor
- [x] Tested
- [x] Clean the code
- [x] Expand with further functionalities
- [x] Generalise

DO NOT MERGE YET, but provide feedback.
[ci skip]

### CODE REVIEW

File: Generator/ClusterSpecificSettings.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
+from copy import deepcopy
+from utils import red
+from functools import total_ordering
+from collections import OrderedDict
+
+@total_ordering
+class TimeFormat():
+    def __init__(self, *args, **kwargs):
+        self.time = {'g':0,'h':0,'m':0,'s':0}
```
Comment (reimersa) [inline Generator/ClusterSpecificSettings.py]: what does `'g'` stand for? Giorno?

File: Generator/ClusterSpecificSettings.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        else:                                  return False
+
+
+
+class ClusterSpecificSettings():
+    def __init__(self):
+        self.Settings = {}
+
+    def getSettings(self, name):
+        if "lxplus" in name.lower():
```
Comment (reimersa) [inline Generator/ClusterSpecificSettings.py:63]: please use `'` as string delimiters, not `"`

File: Generator/ClusterSpecificSettings.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            self.htcondor_lxplus()
+        return self.Settings()
+
+    def slurmGeneral(self):
+        self.Settings["Cluster"] = "slurm"
```
Comment (reimersa) [inline Generator/ClusterSpecificSettings.py:68]: `'` vs `"`

File: Generator/ClusterSpecificSettings.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+    def slurm_PSI(self):
+        self.slurmGeneral()
+        self.Settings['JobFlavour'] = OrderedDict()
+        self.Settings['JobFlavour']['short'] = TimeFormat('1:00:00')
+        self.Settings['JobFlavour']['standard'] = TimeFormat('12:00:00')
+        self.Settings['JobFlavour']['long'] = TimeFormat('24:00:00')
+
+    def htcondorGeneral(self):
+        self.Settings["Cluster"] = "htcondor"
```
Comment (reimersa) [inline Generator/ClusterSpecificSettings.py:78]: `'` vs `"`

File: Generator/ClusterSpecificSettings.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+
+class ClusterSpecificSettings():
+    def __init__(self):
+        self.Settings = {}
+
+    def getSettings(self, name):
+        if "lxplus" in name.lower():
+            self.htcondor_lxplus()
+        return self.Settings()
```
Comment (reimersa) [inline Generator/ClusterSpecificSettings.py:65]: This function only works for lxplus, but not for slurm@psi yet, even though some settings exist already. Why is that?

File: Generator/ClusterSpecificSettings.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        self.Settings['JobFlavour']['nextweek'] = TimeFormat('7-00:00:00')
+
+    def GetTimeUpperLimit(self, givenTime_ = '01:00:00', format='name'):
+        givenTime = TimeFormat(givenTime_)
+        for time in self.Settings['JobFlavour']:
+            if givenTime < self.Settings['JobFlavour'][time]:
+                if format=='name': return time
+                elif format=='time': return self.Settings['JobFlavour'][time]
+                else return (time,self.Settings['JobFlavour'][time])
+        raise ValueError(red('Runtime exceeds 7 days, please choose something a bit faster.'))
```
Comment (reimersa) [inline Generator/ClusterSpecificSettings.py:98]: What is the purpose of this function? It has inconsistent return types, is this intended?

File: Submitter/CondorBase.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+def red(string):
+    return "\x1b[0;31m%s\033[0m"%string
+
+def blue(string):
+    return "\x1b[0;34m%s\033[0m"%string
+
+def bold(string):
+    return "\033[1m%s\033[0m"%strin
+
```
Comment (reimersa) [inline Submitter/CondorBase.py]: why is this repeated? This is defined in `Generator/utils.py`

File: Submitter/CondorBase.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+def prettydic(d, indent=8, color=blue):
```
Comment (reimersa) [inline Submitter/CondorBase.py]: please rename to `prettydict`

File: Submitter/Submitter.py [hunk 9]
```diff
@@ -46,25 +46,47 @@ def Divide(self):
...
-                njobs = len(f.readlines())
+                lines = f.readlines()
+            njobs = len(lines)
             if njobs < 1: continue
-            environ_path = os.getenv('PATH')
-            environ_ld_lib_path = os.getenv('LD_LIBRARY_PATH')
-            command = 'sbatch -a 1-%i -J %s -p %s --chdir %s -t %s submit_analyzer_command.sh %s %s %s' % (njobs, datasetname, queue, os.path.join(self.workdir_local, datasetname, 'joboutput'), runtime_str, missing_files, environ_path, environ_ld_lib_path)
-            jobid = int(subprocess.check_output(command.split(' ')).rstrip('\n').split(' ')[-1])
+
+            if cluster.upper() == "SLURM":
```
Comment (reimersa) [inline Submitter/Submitter.py:65]: `'` vs `"`

File: Submitter/submit.py [hunk 10]
```diff
@@ -17,6 +17,7 @@ def main():
     parser.add_argument('xmlfilename', metavar='XMLFILENAME', type=str, nargs=1, help='Name of the XML file')
     parser.add_argument('--divide', '-d', action='store_true', default=False, dest='divide', help='Divide XMLFILENAME into chunks, create workdir.')
     parser.add_argument('--submit', '-s', action='store_true', default=False, dest='submit', help='Submit split jobs to cluster')
+    parser.add_argument('--cluster', '-k', action='store', type=str, default="htcondor", dest='cluster', help='Submit split jobs to cluster. Default is SLURM')
```
Comment (reimersa) [inline Submitter/submit.py:20]: maybe `-b`, `--batch`?

**Review by reimersa — CHANGES_REQUESTED** [review]

File: Generator/ClusterSpecificSettings.py (hunk 2 above)
Comment (anmalara) [reply Generator/ClusterSpecificSettings.py:63]: done

File: Generator/ClusterSpecificSettings.py (hunk 5 above)
Comment (anmalara) [reply Generator/ClusterSpecificSettings.py:65]: Not complete yet. Will fix it.

File: Submitter/CondorBase.py (hunk 8 above)
Comment (anmalara) [reply Submitter/CondorBase.py]: done

File: Submitter/Submitter.py (hunk 9 above)
Comment (anmalara) [reply Submitter/Submitter.py:65]: done

File: Generator/ClusterSpecificSettings.py (hunk 5 above)
Comment (anmalara) [reply Generator/ClusterSpecificSettings.py:65]: Done

File: Submitter/submit.py (hunk 10 above)
Comment (anmalara) [reply Submitter/submit.py:20]: Not convinced. HTCondor and slurm are clusters. Batch could be a keyword for something else in the future.

File: Generator/ClusterSpecificSettings.py (hunk 6 above)
Comment (anmalara) [reply Generator/ClusterSpecificSettings.py]: Changed. Let's think over this

File: Submitter/UserSpecificSettings.py [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+    def SaveXml(self):
+        with open(self.UserInfo['xml'], 'w') as f:
+            json.dump(self.UserInfo, f, sort_keys=True, indent=4)
+
+    def Set(self,name,info):
+        self.UserInfo.update({name:info})
+
+    def Get(self,name):
+        return self.UserInfo[name]
```
Comment (reimersa) [inline Submitter/UserSpecificSettings.py:30]: it's actually using json, not xml

File: Generator/ClusterSpecificSettings.py [hunk 12]
```diff
@@ -88,11 +94,21 @@ def htcondor_lxplus(self):
...
+        self.htcondorGeneral()
+        # self.Settings['MaxRunTime'] = ('RunTime','00:00:00')
+        # self.Settings['RunTime'] = OrderedDict()
+        # self.Settings['RunTime']['standard'] = TimeFormat('01:00:00')
+        # self.Settings['RunTime']['long'] = TimeFormat('08:00:00')
+
+    def setJobTimeUpperLimit(self, ref_time = '01:00:00'):
+        ref_time_ = TimeFormat(ref_time)
+        if not 'MaxRunTime' in self.Settings:
+            self.Settings['MaxRunTime'] = ('','00:00:00')
```
Comment (reimersa) [inline Generator/ClusterSpecificSettings.py:107]: use `None` as default

File: Generator/ClusterSpecificSettings.py [hunk 13]
```diff
@@ -56,29 +56,35 @@ def __lt__(self, other):
...
         self.slurmGeneral()
-        self.Settings['JobFlavour'] = OrderedDict()
-        self.Settings['JobFlavour']['short'] = TimeFormat('1:00:00')
-        self.Settings['JobFlavour']['standard'] = TimeFormat('12:00:00')
-        self.Settings['JobFlavour']['long'] = TimeFormat('24:00:00')
+        self.Settings['MaxRunTime'] = ('queue','00:00:00')
+        self.Settings['queue'] = OrderedDict()
+        self.Settings['queue']['short'] = TimeFormat('1:00:00')
+        self.Settings['queue']['standard'] = TimeFormat('12:00:00')
+        self.Settings['queue']['long'] = TimeFormat('24:00:00')
```
Comment (reimersa) [inline Generator/ClusterSpecificSettings.py:80]: Here and below: take name of key from first line

File: Submitter/UserSpecificSettings.py [hunk 14]
```diff
@@ -0,0 +1,30 @@
+import os, json
+
+class UserSpecificSettings():
+    """docstring for UserSpecificSettings."""
+
+    def __init__(self, name, email='', cluster=''):
+        self.UserInfo = {
+            'username': name,
+            'xml':      self.GetXMLPath(name),
```
Comment (reimersa) [inline Submitter/UserSpecificSettings.py:9]: info is redundant

File: Submitter/Settings_anmalara.xml [hunk 15]
```diff
@@ -0,0 +1,6 @@
+{
+    "cluster": "htcondor", 
+    "email": "andrea.malara@cern.ch", 
+    "username": "anmalara", 
+    "xml": "/user/anmalara/WorkingArea/LEAF/Submitter/Settings_anmalara.xml"
+}
```
Comment (reimersa) [inline Submitter/Settings_anmalara.xml:6]: remove specific file

File: Submitter/Settings_exampleuser.xml [hunk 16]
```diff
@@ -0,0 +1,6 @@
+{
+    "cluster": "htcondor", 
+    "email": "first.lastname@cern.ch", 
+    "username": "username", 
+    "xml": "SUBMITTERPATH/Settings_username.xml"
```
Comment (reimersa) [inline Submitter/Settings_exampleuser.xml:5]: remove `xml` field

File: Submitter/submit.py [hunk 17]
```diff
@@ -17,12 +17,12 @@ def main():
     parser.add_argument('xmlfilename', metavar='XMLFILENAME', type=str, nargs=1, help='Name of the XML file')
     parser.add_argument('--divide', '-d', action='store_true', default=False, dest='divide', help='Divide XMLFILENAME into chunks, create workdir.')
     parser.add_argument('--submit', '-s', action='store_true', default=False, dest='submit', help='Submit split jobs to cluster')
-    parser.add_argument('--cluster', '-k', action='store', type=str, default="htcondor", dest='cluster', help='Submit split jobs to cluster. Default is SLURM')
+    parser.add_argument('--cluster', '-k', action='store', type=str, dest='cluster', default='', help='Submit split jobs to cluster. Default is taken from UserSpecificSettings')
```
Comment (reimersa) [inline Submitter/submit.py:20]: copy-paste description error

File: Submitter/CondorBase.py [hunk 18]
```diff
@@ -32,46 +38,46 @@ def CreateJobInfo(self, executable="", arguments=""):
...
             'hold': 'True',                                        # to start jobs in hold (eg. debugging)
             # 'transfer_executable':  'True',                        # copy the executable to the cluster Doens't work FIXME
-            'WhenToTransferOutput': 'ON_EXIT_OR_EVICT',                     # specify when to transfer the outout back. Not tested yet
-            # 'requirements':          "OpSysAndVer == 'CentOS7'",   # additional requirements. Not tested yet
+            'WhenToTransferOutput': 'ON_EXIT_OR_EVICT',            # specify when to transfer the outout back. Not tested yet
+            # 'requirements':          'OpSysAndVer == "CentOS7"'',   # additional requirements. Not tested yet
             # '+RequestRuntime':       str(int(nHours*60*60)),       # requested run time. Not tested yet
             }
+        if self.RequestTime!='':
+            self.ModifyJobInfo(self.RequestTime,self.Time)          # Time requested
```
Comment (reimersa) [inline Submitter/CondorBase.py:51]: rename `RequestTime` to a more descriptive name, like `RequestTimeSettingName` or similar

**Review by reimersa — CHANGES_REQUESTED** [review]

File: Submitter/UserSpecificSettings.py (hunk 11 above)
Comment (anmalara) [reply Submitter/UserSpecificSettings.py:30]: Done

File: Generator/ClusterSpecificSettings.py (hunk 12 above)
Comment (anmalara) [reply Generator/ClusterSpecificSettings.py:107]: Done

File: Submitter/UserSpecificSettings.py (hunk 14 above)
Comment (anmalara) [reply Submitter/UserSpecificSettings.py:9]: done

File: Submitter/Settings_anmalara.xml (hunk 15 above)
Comment (anmalara) [reply Submitter/Settings_anmalara.xml:6]: Done

File: Submitter/Settings_exampleuser.xml (hunk 16 above)
Comment (anmalara) [reply Submitter/Settings_exampleuser.xml:5]: Done

Comment (anmalara) [reply Submitter/Settings_exampleuser.xml:5]: Done

File: Submitter/CondorBase.py (hunk 18 above)
Comment (anmalara) [reply Submitter/CondorBase.py:51]: Done

File: Submitter/submit.py (hunk 17 above)
Comment (anmalara) [reply Submitter/submit.py:20]: Done

Comment (anmalara) [reply Submitter/submit.py]: Done

**Review by reimersa — APPROVED** [review]

**Pull request merged**